### ISSUE TITLE
Strict conditional type checking

### ISSUE THREAD

**Issue by tkaden4** [microsoft/TypeScript#42560]
# Suggestion

## 🔍 Search Terms

strict, boolean, if, warning, sound, type

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

My suggestion is around the soundness of `if` statements in TypeScript code when the condition is not typed as a `boolean`. I think that the ability to have non-booleans as an argument to the `if` construct is acceptable as a Javascript compatibility feature, but for well-typed code it is misguided. I believe there should be an added optional compiler flag (default false) flag to turn on stricter conditional checking, restricting the type of the arguments to ternary expressions, `if` statements, and `else if` statements to be `boolean`, instead of any value, as it is now. I propose the name of this flag to be something akin to `strictConditionals` or `safeConditionals`.

## 📃 Motivating Example

I've often run into the following problem while using async functions in TypeScript:
```typescript
async function returnsABoolean(): Promise<boolean> {
  // ...
}
// ...
if(returnsABoolean()) {
  // then ...
} else {
  // else ...
}
``` 

A simple bug to fix, but one that seems like low-hanging fruit for the type checker to handle. The problem is that a `Promise` value is considered truthy here, but what we are really trying to do is this:
```typescript
async function returnsABoolean(): Promise<boolean> {
  // ...
}
// ...
if(await returnsABoolean()) {
  // then ...
} else {
  // else ...
}
``` 
In the first case, the `if` branch will always execute, whereas in the second case, the branch that gets executed is purely dependent on the return value of `returnsABoolean()`.

The crux of the problem is that conditionals are not restricted to `boolean` conditions, but rather any value that is then evaluated as `true`/`false` according to Javascript's truthiness rules, which can often be unintuitive.

## 💻 Use Cases

I would use this feature to further improve the type-safety of my current code and reduce the # of bugs that get produced because of the awkward truthiness rules that Javascript uses.

One workaround is to define a function called `safeIf` that takes in thunks, like so:
```typescript
export function safeIf<T>(cond: boolean, ifB: () => T, elseB: () => T): T {
  return cond ? ifB() : elseB();
}
```
This function implements the same functionality as a ternary expression, but is strict on the type of the condition. Following the previous example, we would have this:
```typescript
async function returnsABoolean(): Promise<boolean> {
  // ...
}
// ...
safeIf(returnsABoolean(), () => {
  // then ...
}, () => {
  // else ...
})
```
which successfully errors while type-checking, as `returnsABoolean()` is of type `Promise<boolean>` and not `boolean`. However, this is a cumbersome and non-idiomatic way to write TypeScript/Javascript code, hence the feature request.

**Comment by MartinJohns** [comment]
There's already an eslint rule existing for this: [`strict-boolean-expressions`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md)
